### PR TITLE
Allow collection size constraint

### PR DESCRIPTION
**Pull request by chakhal**
Hello,

There is a contribution that aims to add size limitations on the field collection (I needed that for a project I'm working on).
This is my first contribution \o/
So if you're interested in, please, comment on whatever you want:
- coding style
- options/event naming strategy
- what about a global configuration that could be override at field level (only field level is implemented right now)

Below, my commit description:

Allow to define collection size constraint via data attribute on the field
- data-constrained-collection-min-size
- data-constrained-collection-max-size

if min size is defined, then, at initialization, at least [min size items] are created (if not already existing)
if max size is defined, it is not possible to add more items when max size is reached

New events triggered on this._container:
- prototype.maxsize-reached (when the last permitted item is added)
- prototype.maxsize-already-reached (when addField is called but the maximum size has already been reached)

### CODE REVIEW

**Comment by wiese** [comment]
Hi chakhal,
congrats for joining the open source world, the guts to issue a pull request, and the compliment of contributing to one of my projects.
I like the idea, it is definitely a useful feature, but there are a couple of things I would like you to lay focus on, before we can get this contribution into the main project (if you have the time to):
- I am not sure how these limits (they definitely exist) are implemented in symfony - the plugin in meant for use with sf, so we have to keep it compatible with it. Do 'data.constrainedCollectionMinSize' and max exist in sf, how are they called, how are they passed to the view? We want to get the data from the default form generator/validator without having to fiddele (we can still do if we chose to, but not as a requirement) with the view ourselves
- This project uses grunt, includes linting and tests (qunit). The pull requests unforutnatly breaks the travis build (CI system), because it violates the linting. Furthermore your commits do not provide additional tests, while heavily modifying the core of the project. https://travis-ci.org/wiese/jquery-sfPrototypeMan/builds/41618348 This is a no-go. Tests must be provided to prove both, that the previously existing functionality still works (some are in place from my contributions), as well as proving that your contributions work, too, when enabled via config. These automates tests (run 'grunt ci') would immensely help you to polish out edge cases, give you confidence for changes, and boost the reviewer's ability to understand what is acutally going on (like a demo) and merge them - without them I will not merge it, as it would potentially introduct bugs, and generally lower the software quality of the project: the opposite of what we want.
- If the config changes, the documentation should, too! Document the newly available config options in both the api documentation (e.g. README) as well as the demo. 

Sorry, if that sounds harsh, but I also hope that by giving you this input you can improve your contribution skills (also for other projects) and make you become a valueable contributor to the open source community. Don't get discouraged, we all had got to hear that at some point - just keep working it and it will become second nature to you, as it actually makes life easier.

As a general rule of thumb, let's try to stick to these contribution rules: 
- http://contribute.jquery.org/
- http://symfony.com/doc/current/contributing/code/patches.html#work-on-your-patch
- tests, tests, tests (maybe even extening tests for the core features before adding, changes to be safe)

Cheers
wiese

**Comment by chakhal** [comment]
Thank you for your feedback.
I will have a look at your links and commit some tests!

To answer you, in my case, the data attributes are handled by a symfony custom field extension that I wrote for a projet I'm working on.
(Extension of CollectionType that handles options min_size and max_size).

Another question about the process:
If I can write some tests, do you want me to close this pull request and open another one ?

chakhal

**Comment by wiese** [comment]
Hi chakhal,

if you can still add further commits to this pull request later on (not sure), I don't mind keeping it open - otherwise just close it, and create another (less formal) one when you found the time to amend the changes.

Have not had a use case for limiting the number of elements inside a collection myself, but are you aware of http://symfony.com/doc/current/reference/constraints/Count.html ?

Cheers
wiese

File: jquery.sfprototypeman.js [hunk 1]
```diff
@@ -30,6 +30,7 @@
 		$(config.containerSelector, context).each(function() {
 			var container = new SfPrototypeContainer($(this), config);
 			container.applyExtensions();
+            container.applySizeConstraints();
```
Comment (wiese) [inline jquery.sfprototypeman.js:33]: indention using tabs, please
http://contribute.jquery.org/style-guide/js/#spacing

File: jquery.sfprototypeman.js [hunk 2]
```diff
@@ -61,6 +63,23 @@
 			this._extendFields();
 		},
 
+		applySizeConstraints: function() {
+			var data = this._container.data(),
+				counter = this._getExisting().length;
+			if (data.constrainedCollectionMinSize) {
```
Comment (wiese) [inline jquery.sfprototypeman.js:69]: use mapping of data-attributes to config keys (or name them identically), specify them once in an array and implement the "copying" from html programmatically (loop), not individually - so it scales for more options to come

File: jquery.sfprototypeman.js [hunk 3]
```diff
@@ -89,10 +108,26 @@
...
 		addField: function() {
 			var counter = this._getExisting().length,
-				newElement = this._createField(counter);
-			newElement.appendTo(this._container);
+				maxSize = this._fieldConfig.maxSize,
+				mustAppend = true;
+
+			if (typeof maxSize != "undefined") {
+				if (counter >= maxSize){
+					mustAppend = false;
```
Comment (wiese) [inline jquery.sfprototypeman.js:116]: avoid unneeded variables. just return, hence avoid the code blocks to come to be nested in conditionals
https://en.wikipedia.org/wiki/Cyclomatic_complexity

File: jquery.sfprototypeman.js [hunk 4]
```diff
@@ -89,10 +108,26 @@
...
+
+			if (mustAppend){
+				var newElement = this._createField(counter);
+				newElement.appendTo(this._container);
 
-			this._container.trigger("prototype.added", [this]);
+				this._container.trigger("prototype.added", [this]);
+
+				if (maxSize && counter - maxSize === 1){
+					this._container.trigger("prototype.maxsize-reached", [this]);
```
Comment (wiese) [inline jquery.sfprototypeman.js:128]: I would limit it to the event fired on refused adding, and scrap this one - but thats a matter of personal taste. Reason: different behaviour (no such event) on initialization if all the fields come from the view.

File: jquery.sfprototypeman.js [hunk 5]
```diff
@@ -89,10 +108,26 @@
...
 			var counter = this._getExisting().length,
-				newElement = this._createField(counter);
-			newElement.appendTo(this._container);
+				maxSize = this._fieldConfig.maxSize,
+				mustAppend = true;
+
+			if (typeof maxSize != "undefined") {
+				if (counter >= maxSize){
+					mustAppend = false;
+					this._container.trigger("prototype.maxsize-already-reached", [this]);
```
Comment (wiese) [inline jquery.sfprototypeman.js:117]: If there is a maxsize event, there should probably be a minsize event in _rmButtonCallback, right?!

File: jquery.sfprototypeman.js [hunk 6]
```diff
@@ -41,6 +42,7 @@
 	SfPrototypeContainer = function(container, config) {
 		this._container = container;
 		this._config = config;
+        this._fieldConfig = {};
```
Comment (wiese) [inline jquery.sfprototypeman.js:45]: why use an extra variable? if you fear for polluting the namespace in the config, you can use one object with again multiple members. This way you have the freedom to either let the limits be pulled from data- or pass it as settings in constructor (programmatically)

**Comment by wiese** [comment]
Closed due to inactivity
